tracer/src/Datadog.Trace/TraceProcessors/TruncatorTraceProcessor.cs

```diff
@@ -0,0 +1,98 @@
+// </copyright>
+
+using System;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.TraceProcessors
+{
+    internal class TruncatorTraceProcessor : ITraceProcessor
+    {
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/truncator.go
+
+        // Values from: https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/truncate.go#L21-L28
+        // MaxResourceLen the maximum length a span resource can have
+        internal const int MaxResourceLen = 5000;
+        // MaxMetaKeyLen the maximum length of metadata key
+        internal const int MaxMetaKeyLen = 200;
+        // MaxMetaValLen the maximum length of metadata value
+        internal const int MaxMetaValLen = 25000;
+        // MaxMetricsKeyLen the maximum length of a metric name key
+        internal const int MaxMetricsKeyLen = MaxMetaKeyLen;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<TruncatorTraceProcessor>();
+        private readonly TruncatorTagsProcessor truncatorTagsProcessor = new TruncatorTagsProcessor();
+
+        public TruncatorTraceProcessor()
+        {
+            Log.Information("TruncatorTraceProcessor initialized.");
+        }
+
+        public ArraySegment<Span> Process(ArraySegment<Span> trace)
+        {
+            for (var i = trace.Offset; i < trace.Count + trace.Offset; i++)
+            {
+                trace.Array[i] = Process(trace.Array[i]);
+            }
+
+            return trace;
+        }
+
+        public Span Process(Span span)
+        {
+            // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/truncator.go#L17-L21
+            span.ResourceName = TruncateResource(span.ResourceName);
+
+            // Set the tags processor
+            if (span.Tags is TagsList tagsList)
+            {
+                tagsList.AddTagProcessor(truncatorTagsProcessor);
+            }
+
+            return span;
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/truncate.go#L30-L34
+        internal static string TruncateResource(string r)
+        {
+            if (TraceUtil.TruncateUTF8(ref r, MaxResourceLen))
+            {
+                Log.Debug("span.truncate: truncated `Resource` (max {maxResourceLen} chars): {resource}", MaxResourceLen, r);
+            }
+
+            return r;
+        }
+
+        private class TruncatorTagsProcessor : ITagProcessor
+        {
+            // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/truncator.go#L26-L44
+            public void ProcessMeta(ref string key, ref string value)
+            {
+                if (TraceUtil.TruncateUTF8(ref key, MaxMetaKeyLen))
+                {
+                    key += "...";
```

Comment:
If we're adding three characters subsequently, should this be `MaxMetaKeyLen - 3`? Also, should this definitely be `...` (3x`.`) or `…`(`&hellip`)

Reply:
Well we are doing the same as the agent: https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/truncator.go#L32 so I guess is ok.